pkg/virt-handler/cmd-client/client.go

```diff
@@ -103,177 +103,171 @@ func SocketFromUID(baseDir string, uid string) string {
 	return filepath.Join(SocketsDirectory(baseDir), sockFile)
 }
 
-func GetClient(socketPath string) (LauncherClient, error) {
-	conn, err := rpc.Dial("unix", socketPath)
+func NewClient(socketPath string) (LauncherClient, error) {
+	conn, err := grpcutil.DialSocket(socketPath)
 	if err != nil {
+		log.Log.Reason(err).Infof("Failed to dial notify socket: %s", socketPath)
 		return nil, err
 	}
+	client := cmdv1.NewCmdClient(conn)
 
-	return &VirtLauncherClient{client: conn}, nil
+	return &VirtLauncherClient{
+		client: client,
+		conn:   conn,
+	}, nil
 }
 
 func (c *VirtLauncherClient) Close() {
-	c.client.Close()
+	c.conn.Close()
 }
 
-func (c *VirtLauncherClient) genericSendCmd(args *Args, cmd string) (*Reply, error) {
-	reply := &Reply{}
+func (c *VirtLauncherClient) genericSendVMICmd(cmdName string,
+	cmdFunc func(ctx context.Context, request *cmdv1.VMIRequest, opts ...grpc.CallOption) (*cmdv1.Response, error),
```

Comment:
VMI versioning here is tricky. Somehow we need to make sure we're only sending a version of the VMI that the server knows about or is compatible with. 

Essentially I want cmdv1 client to explicitly break when we break the VMI api compatiblity. 

What would you think about putting a check here in this function that rejects any VMI that isn't of version v1alpha3? In the future as we progress our api, this command will break forcing us to make a decision about whether to allow versions other that v1alpha3 to be communicated here or not. 

Eventually I could see a vmi version compatibility list for cmdv1 where something like [v1alpha3, v1beta1, v1] would be compatible. If the vmi api ever breaks in an incompatible way, we'd have to create a cmdv2 to handle that I suppose. 

thoughts?

Reply:
I think with my suggestion at another place, about using constants for supportedVersions for the info server, and strings in the version check for the info client, this is solved.